### PR TITLE
[Rule] Keep last events with repository

### PR DESCRIPTION
**Pull request by tjoyal**
Alternative on https://github.com/christianblais/triggerify/pull/113 with `RuleEventRepository`

Keep the last 10 events on each rules

### CODE REVIEW

File: lib/rule_runner.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    rule_event.add_detail(:info, "Event received")
+
+    rule.filters.each.with_index do |filter, index|
+      if FilterValidator.new(filter).valid?(callback)
+        rule_event.add_detail(:info, "Filter ##{index + 1}: Met")
+      else
+        rule_event.add_detail(:info, "Filter ##{index + 1}: Unmet, event dropped")
+        return
+      end
+    end
```
Comment (christianblais) [inline lib/rule_runner.rb:17]: Do we assume filters are always ordered, in the same way, or should we capture the return value of all filters, log it, and then return if one of them is invalid?

File: app/models/rule.rb [hunk 2]
```diff
@@ -53,8 +53,18 @@ class Rule < ActiveRecord::Base
   validates :name, presence: true
   validates :topic, presence: true, inclusion: TOPICS.keys
 
+  serialize :events, RuleEventRepositorySerializer
+
+  after_initialize :default_rule_event_repository
```
Comment (christianblais) [inline app/models/rule.rb:58]: Why is this required? Wouldn't empty events still return a repository given `      return RuleEventRepository.new if string.nil?`?

File: lib/rule_runner.rb (hunk 1 above)
Comment (tjoyal) [reply lib/rule_runner.rb:17]: So 2 questions here:

> Do we assume filters are always ordered

Let's throw in an `order(:id)` to guarantee the behaviour.

For now I settled with "Filter #X" but we should revisit this to see if it can be made better visually. 

> Should we capture the return value of all filters, log it, and then return if one of them is invalid?

Why not. I can do this in this PR if it is what you would like.

File: app/models/rule.rb (hunk 2 above)
Comment (tjoyal) [reply app/models/rule.rb:58]: You are right! I didn't know serializers were run on initialize

File: lib/rule_runner.rb (hunk 1 above)
Comment (tjoyal) [reply lib/rule_runner.rb:18]: Updated to run all filters and log the result of each (will still stop the process if any fail)

File: db/migrate/20220226134505_add_events_to_rules.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddEventsToRules < ActiveRecord::Migration[6.1]
+  def change
+    add_column :rules, :events, :text
```
Comment (christianblais) [inline db/migrate/20220226134505_add_events_to_rules.rb:3]: Wait, if all you want is JSON instead of ruby marshalled objects, shouldn't this be a JSON column? This would allow us to avoid having to compare to `nil` and would give use automatic parse/dump?

Comment (tjoyal) [reply db/migrate/20220226134505_add_events_to_rules.rb:3]: >  shouldn't this be a JSON column?

Yes

> This would allow us to avoid having to compare to nil

Only if we add the default value in mysql schema which I don't really agree with.

> would give use automatic parse/dump?

No. The JSON structure(as for text) does not know the object whcih needs to be re-hydrated. This is why vanilla rails serialization marshal dump the object data and class

File: lib/rule_event_repository_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+      repository.dump
+    end
+
+    def load(data)
+      return RuleEventRepository.new if data.nil?
+
+      RuleEventRepository.load(data)
+    end
+  end
+end
```
Comment (tjoyal) [inline lib/rule_event_repository_serializer.rb:13]: With the JSON database format, this is now much simpler.

One one hand, serialization != business logic. On the other hand, I could see `RuleEventRepositorySerializer` logic being folded in `RuleEventRepository` and to use `serialize :events, RuleEventRepository`.

While verbose, and since already coded, I think it make sense to keep them separate. If you prefer I can make the change. Since this is only ruby and not related to the data model, I don't really mind either way.

File: lib/rule_event_repository_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class RuleEventRepositorySerializer
+  class << self
+    def dump(repository)
+      repository.dump
+    end
+
+    def load(data)
+      return RuleEventRepository.new if data.nil?
```
Comment (tjoyal) [inline lib/rule_event_repository_serializer.rb:8]: This is needed unless we do `t.json "events", default: '{"events": []}'` which would bleed ruby logic in mysql land.

Comment (christianblais) [reply lib/rule_event_repository_serializer.rb:8]: So let's do `default: '[]'`?

Comment (tjoyal) [reply lib/rule_event_repository_serializer.rb:8]: We don't "need" it, but it is a nice separation of concerns.

> While verbose, and since already coded, I think it make sense to keep them separate. If you prefer I can make the change. Since this is only ruby and not related to the data model, I don't really mind either way.

I can add a commit that removes it and we'll see

Comment (tjoyal) [reply lib/rule_event_repository_serializer.rb]: Commit that removes the serializer: https://github.com/christianblais/triggerify/pull/116/commits/8167a129ce525b789bc2d8e6ed80be95c8f32bdb
